mantidproject/mantid
# Support saving a multi-dataset ORSO file from the ISIS Reflectometry GUI

### Description of work

#### Summary of work

This PR updates the ISIS Reflectometry GUI to provide options for saving multiple datasets to a single ORSO file. The functionality actually performing the save was added to the `SaveISISReflectometryORSO` algorithm in PR #37157.



Fixes #36396. 




#### Further detail of work


- Checkbox `Save multiple datasets to a single file` has been added to allow users to choose whether to combine multiple datasets into a single file when auto-saving into the ORSO format. When an ORSO file is saved out that contains more than one dataset the suffix `_multi` is appended to the filename. As agreed with our scientists, the multi-dataset file will be limited to auto-saving in this first version - if workspaces are manually selected and saved via the Save ASCII tab they will go into separate ORSO files for now.

- A second new checkbox, `Additional columns (includes Q resolution)`, has been added to give the option to include additional data columns when saving non-stitched datasets into the ORSO format. Following discussion with our scientists, this PR also changes the default behaviour if for some reason the additional columns have been requested and cannot be calculated (an edge case that shouldn't be encountered via the GUI). In this case, a warning will now be logged and the file will be saved without them.

- I have included a couple of small fixes and tests for a couple of places where you could cause a crash when trying to perform actions in the save tab with names of workspaces that were no longer in the GUI.

- Relevant documentation has been updated.

- On the linked issue there is a final bullet point that has not been done in this PR (adding a right click option from the Runs tab batch table to manually save a group into a multi-dataset file). I will be opening a separate issue to look at this.

### To test:



1) Go to `Interfaces -> Reflectometry -> ISIS Reflectometry`.
1) Go to the Save ASCII tab. Set a Save Path for saving the file to.
1) In the File Format section, select `ORSO ASCII (*.ort)` from the drop-down.
1) In the Automatic Save section, tick the `Save as ASCII on completion` checkbox.
1) The `Include individual row outputs for groups` checkbox should be enabled. Select this checkbox.
1) The `Save multiple datasets to a single file` checkbox should be enabled and selected by default.
1) Go to the Runs tab, enter Investigation ID `1120015` and Cycle `11_3`. Click the Search button.
1) From the search results, click to highlight runs `11934`, `11935` and `11936` and click the Transfer button to transfer them into the batch table on the right hand side. Click the Process button to reduce the data.
1) If you navigate to the save path that you set in the Save ASCII tab, you should see a single `.ort` file with the suffix `_multi` in the filename. This file should contain multiple datasets. Assuming you have left the `Additional columns (includes Q resolution)` checkbox ticked, the non-stitched (individual angle) datasets should have 8 columns and the final stitched dataset should have 4 columns.
1) Go back to the Save ASCII tab and from the List of Workspaces section on the left, select workspaces `IvsQ_11934_11935+11936` and  `IvsQ_binned_11934`.
1) Click the big Save button on the right hand side to manually save to file. If you navigate to the save path that you set, you should see two separate `.ort` files with one dataset in each. This tests that the manual save is still outputting multiple datasets to separate files.
1) Still on the Save ASCII tab, in the Automatic Save section untick the `Include individual row outputs for groups` checkbox. This means that only the stitched dataset from a group multi-row reduction will be automatically saved out.
1) Go back to the Runs tab and re-process the group you just reduced.
1) If you navigate to the save path that you set in the Save ASCII tab you should see a single `.ort` file, but this time it should **not** have the suffix `_multi` in the filename. This file should contain only the stitched dataset and should have just 4 columns.
1) Test that the checkboxes on the Save ASCII tab are enabled and disabled correctly depending on the selected file format and auto-save options. The `Additional columns (includes Q resolution)` checkbox should only be enabled for the ORSO format. The `Save multiple datasets to a single file` checkbox should only be enabled if `Save as ASCII on completion` has been ticked and the ORSO file format has been selected. It should be disabled otherwise.





---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
